### DISCUSSION TITLE
Issue with auth refresh in node.js server

### DISCUSSION THREAD

**Discussion by Kevand** [pocketbase/pocketbase#2071] [Q&A] [answered]
Hi!,

I've got a problem with auth refresh function.
I'm trying to authenticate a user before he uses the API, but whenever there is a request, i'm getting
`ClientResponseError 404: Not Found.` from authRefresh function.

Here is my code:
```typescript
const auth = async (req: Request, res: Response, next: NextFunction) => {
  req.pbClient = new PocketBase('https://myurltopocketbase.com');

  const token = req.get("Authorization") || "";

  req.pbClient.authStore.loadFromCookie(token);

  req.pbClient.authStore.onChange(() => {
    res.setHeader('Set-Cookie', req.pbClient.authStore.exportToCookie({httpOnly: false}));
  });

  try{
    if(req.pbClient.authStore.isValid){
        await req.pbClient.collection('users').authRefresh();
    }
  } catch (e) {
    console.log(e);
    req.pbClient.authStore.clear();
    return res.sendStatus(401);
  }

  next();
}
```

**Comment by ganigeorgiev** [comment, marked as answer]
I think this is because the below part

```js
  const token = req.get("Authorization") || "";

  req.pbClient.authStore.loadFromCookie(token);
```

is incorrect.

If you want to load a plain token value, you need to use `req.pbClient.authStore.save(token, null)`.

If you want to load a cookie, then the cookie must be a valid cookie string, eg. `req.headers.get('cookie')`.

**Reply by Kevand** [reply]
Alright, but i doesn't work for me either when i use cookie.
I'm not sure if i'm correctly sending a request with it, is it automaticly added to fetch request or do i need to add it somehow?

**Reply by ganigeorgiev** [reply]
I'm confused. How is the above node endpoint triggered? Is this part of a SSR setup?

**Reply by Kevand** [reply]
Sorry for confusion, It's not SSR, It is triggered by fetch request from normal Svelte app

**Reply by ganigeorgiev** [reply]
There is not enough context to understand what you are trying to do.

How do you interact with PocketBase - is only the node server sending requests to PocketBase or both client and node can send requests to PocketBase?

Are you manually sending the `Authorization` header with the fetch request to the node server?

Please provide a more detailed code sample or at least elaborate what is your setup.

**Reply by Kevand** [reply]
The way it works is:
1. Both client and server have access to pocketbase
2. Client uses it to login and access data from databases
3. Node server is used for file storage (I wanted more freedom when it comes to file size, names etc...)
4. When client login in it sends request to node server to get list of files
5. And at this point i've got an error i mentioned at the beginning
The thing i want is to give access to files only for authorised clients

**Reply by ganigeorgiev** [reply]
@Kevand  Please provide a more detailed code sample or minimal reproducible repo because there are way too many external factors and there is not enough information to deduce what exactly is happening and what requests you are sending and how.

If you have a mixed access approach (aka. both client and node can send requests), then you need somehow to propagate the auth state from one env to the other. This could happen 2 ways:

1. Manually send the `pb.authStore.token` in a `Authorization` header with your custom fetch request to the node server, then read the header value and save it in the request SDK instance with `pb.authStore.save(token, null)`.

2. Use cookies. This also allow the node server to be able to modify the auth state. For this approach you have to load and set `document.cookie` from the client:
    ```js
    pb.authStore.loadFromCookie(document.cookie);
    pb.authStore.onChange((token, model) => {
        document.cookie = pb.authStore.exportToCookie({ httpOnly: false })
    });
    ```
   **Note that PocketBase API is completely stateless.** A request is considered authenticated as long it has `Authorization` header with it. This header is automatically added when making requests with the SDK if there is a nonexpired `pb.authStore.token`. Cookies are used just for "sharing" the auth data between the client and node.

**Reply by Kevand** [reply]
Alright, i don't think i need to add anything more. Thank you so much for help and overall explanation :)
